src/main/java/techreborn/enet/ElectricNetwork.java

```diff
@@ -0,0 +1,238 @@
+import net.minecraft.util.math.Direction;
+import team.reborn.energy.EnergyTier;
+import techreborn.TechReborn;
+import techreborn.blockentity.cable.CableBlockEntity;
+
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public final class ElectricNetwork {
+	private final EnergyTier networkTier;
+	private final HashSet<CableBlockEntity> cableBlockEntities = new HashSet<>();
+	private boolean isDirty;
+	private boolean isEnergized;
+
+	public ElectricNetwork(EnergyTier tier) {
+		networkTier = tier;
+		isDirty = true;
+	}
+
+	public boolean isEmpty() {
+		return cableBlockEntities.isEmpty();
+	}
+
+	public void addBlockEntity(CableBlockEntity entity) {
+		isDirty = cableBlockEntities.add(entity) || isDirty;
+	}
+
+	public void removeBlockEntity(CableBlockEntity entity) {
+		isDirty = cableBlockEntities.remove(entity) || isDirty;
+	}
+
+	public void tick() {
+		if (isEmpty()) {
+			return;
+		}
+
+		if (isDirty) {
+			TechReborn.LOGGER.debug("Electric Network {} is dirty, checking for network and connection changes", this);
+			walkNetwork();
+
+			isDirty = false;
+		}
+
+		distributePower();
+	}
+
+	public boolean isEnergized() {
+		return isEnergized;
+	}
+
+	private void walkNetwork() {
+		// If the network is empty, why bother?
+		if (cableBlockEntities.isEmpty()) {
+			return;
+		}
+
+		TechReborn.LOGGER.debug("Walking the Electric Network to detect for changes in neighbors...");
+
+		// Keep track of the nodes we've seen and visited
+		HashSet<CableBlockEntity> seenNodes = new HashSet<>();
+		HashSet<CableBlockEntity> visitedNodes = new HashSet<>();
+
+		// "See" the first node in the list
+		CableBlockEntity firstNode = cableBlockEntities.iterator().next();
+		TechReborn.LOGGER.debug("First node in the walk: {}", firstNode);
+		seenNodes.addAll(visitCableNode(firstNode, visitedNodes));
+
+		// "remainingNodesToVisit" is our seen nodes, excluding the ones we've visited
+		HashSet<CableBlockEntity> remainingNodesToVisit = new HashSet<>(seenNodes);
+
+		while (!remainingNodesToVisit.isEmpty()) {
+			for (CableBlockEntity nodeToVisit : remainingNodesToVisit) {
+				seenNodes.addAll(visitCableNode(nodeToVisit, visitedNodes));
+			}
+
+			remainingNodesToVisit.clear();
+			remainingNodesToVisit.addAll(seenNodes);
+			remainingNodesToVisit.removeAll(visitedNodes);
+		}
+
+		for (CableBlockEntity nodeConnected : seenNodes) {
+			if (nodeConnected.getElectricNetwork() != this) {
+				TechReborn.LOGGER.debug(
+						"Block {} network change: {} -> {}",
+						nodeConnected.getPos().asLong(),
+						nodeConnected.getElectricNetwork(),
+						this);
+
+				nodeConnected.setElectricNetwork(this);
+			}
+		}
+
+		TechReborn.LOGGER.debug(
+				"Starting purge of disconnected entities. Have {}, seen {}",
+				cableBlockEntities.size(),
+				seenNodes.size());
+
+		cableBlockEntities
+				.stream()
+				.filter(node -> !seenNodes.contains(node))
+				.collect(Collectors.toList())
```

Comment:
Why this collect call?

Reply:
This is needed because the `cableBlockEntities` collection can change when walking the network, so we need to materialize the results to avoid a concurrent modification exception.